llama.cpp

```diff
@@ -5862,37 +5902,76 @@ static void tokenizer_st_partition(const llama_vocab & vocab, std::forward_list<
                     
 
                     // right
-                    if (match + special_token.length() < raw_text->length())
+                    if (match + special_token.length() < raw_text_base_offset + raw_text_base_length)
                     {
-                        buffer.emplace_after(it, raw_text->substr(match + special_token.length()));
+                        /*
+                                        |                            |
+                        -------------------------------------------------------------------------
+                                        .       |ttttt|              |
+                        */
+                        //buffer.emplace_after(it, raw_text->substr(match + special_token.length()));
+                        const int64_t right_reminder_offset = match + special_token.length();
+                        const int64_t right_reminder_length = raw_text_base_length - ( ( match - raw_text_base_offset ) + special_token.length() );
+                        buffer.emplace_after(it, (*raw_text), right_reminder_offset, right_reminder_length);
+
+#ifdef PRETOKENIZERDEBUG
+                        fprintf(stderr,"FR: (%ld %ld) '%s'\n", right_reminder_offset, right_reminder_length, raw_text->substr(right_reminder_offset, right_reminder_length).c_str());
+#endif
+
                         it++;
 
                         if (source == 0)
                         {
+                            // TODO? It might not be needed to store/restore the iterator like this
+                            //  but this gives me the peace of mind I'm not causing some
+                            //  accidental undefined behaviour.
+                            auto it_backup = std::distance( buffer.begin(), it );
+
                             buffer.erase_after(buffer.before_begin());
```

Comment:
https://en.cppreference.com/w/cpp/container/forward_list
> Adding, removing and moving the elements within the list, or across several lists, does not invalidate the iterators currently referring to other elements in the list. However, an iterator or reference referring to an element is invalidated when the corresponding element is removed (via [erase_after](https://en.cppreference.com/w/cpp/container/forward_list/erase_after)) from the list.

Reply:
Thank you, I'll remove it then.